src/lib/core/common-behaviors/initialized.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Constructor} from './constructor';
+import {Observable} from 'rxjs/Observable';
+import {Subscriber} from 'rxjs/Subscriber';
+
+/**
+ * Mixin that adds an initialized property to a directive which, when subscribed to, will emit a
+ * value once markInitialized has been called, which should be done during the ngOnInit function.
+ * If the subscription is made after it has already been marked as initialized, then it will trigger
+ * an emit immediately.
+ * @docs-private
+ */
+export interface OnInitialized {
+  /** Stream that emits once during the directive/component's ngOnInit. */
+  initialized: Observable<void>;
+
+  /**
+   * Sets the state as initialized and must be called during ngOnInit to notify subscribers that
+   * the directive has been initialized.
+   * @docs-private
+   */
+  _markInitialized: () => void;
+}
+
+/** Mixin to augment a directive with an initialized property that will emits when ngOnInit ends. */
+export function mixinInitialized<T extends Constructor<{}>>(base: T):
+    Constructor<OnInitialized> & T {
+  return class extends base {
+    /** Whether this directive has been marked as initialized. */
+    _isInitialized = false;
+
+    /**
+     * List of subscribers that subscribed before the directive was initialized. Should be notified
+     * during _markInitialized.
+     */
+    _pendingSubscribers: Subscriber<void>[] = [];
+
+    /**
+     * Observable stream that emits when the directive initializes. If already initialized, the
+     * subscriber is stored to be notified once _markInitialized is called.
+     */
+    initialized = new Observable<void>(subscriber => {
+      // If initialized, immediately notify the subscriber. Otherwise store the subscriber to notify
+      // when _markInitialized is called.
+      if (this._isInitialized) {
+        this._notifySubscriber(subscriber);
+      } else {
+        this._pendingSubscribers.push(subscriber);
+      }
+    });
+
+    constructor(...args: any[]) { super(...args); }
+
+    /**
+     * Marks the state as initialized and notifies pending subscribers. Should be called at the end
+     * of ngOnInit.
+     * @docs-private
+     */
+    _markInitialized(): void {
```

Comment:
Should this guard against being called twice?

Reply:
The logic in the method is safe if called twice, but to make it more rigid I've added an error that makes it clear when it is called twice. I suspect it'll be raised in cases where a subclass tries to call it even though a base class does so. Might prevent some erroneous calls.